### DISCUSSION TITLE
How to model Covariance matrices?

### DISCUSSION THREAD

**Discussion by jan-glx** [TuringLang/Turing.jl#1526] [Q&A]
I am trying to do inference in the following simple model:

```
using Optim
using StatsPlots
using Turing
using LinearAlgebra


D = 8
N = 100
Ψ =  Matrix{Float64}(I * 10.0, D, D)

# Define a simple Normal model with unknown mean and variance.
@model function mvn_model(X, ::Type{T} = Float64) where {T}
  if X === missing
    # Initialize `x` if missing
    X = Matrix{T}(undef, N, D)
  end
  σ = Vector{T}(undef, D)
  μ ~ MvNormal(zeros(Float64, D), Matrix{Float64}(I * 10.0, D, D))
  Ρ ~ LKJ(D, 2.0)
  for d in 1:D
    σ[d] ~ truncated(Normal(0.0, 1.0), 0.0, Inf)
  end
  Σ = diagm(σ) * Ρ * diagm(σ)
  for n in 1:N
      X[n, :] ~ MvNormal(μ, Σ)
  end
  return (X, μ, Σ)
end
example_data = mvn_model(missing)()
#  Run sampler, collect results
chn1 = sample(mvn_model(example_data[1]), HMC(0.1, 5), 1000)
chn2 = sample(mvn_model(example_data[1]), NUTS(0.65), 1000)
```
but it fails with `PosDefException: matrix is not Hermitian; Cholesky factorization failed.`

`Stan` offers `lkj_corr_cholesky` and `multi_normal_cholesky` to prevent this problem and enable more efficient sampling. Is something like this possible with `Turing.jl` as well?

Let me know if there is a better place to ask this!

**Comment by just-cameron** [comment]
The trick I've done for this in the past is to enforce symmetry for all the covariance matrices with `Symmetric(thing)`. In your case I suspect the problem matrix should be

```julia
Σ = Symmetric(diagm(σ) * Ρ * diagm(σ))
```

You have another covariance matrix there too, and it's hard to tell which is causing the issue without line numbers, but you can enforce symmetry there with the same principle.

**Reply by jan-glx** [reply]
Thanks for the quick reply! Unfortunately, the error is thrown  from the `Ρ ~ LKJ(D, 2.0)` statement (and your suggestion does not fix *this* problem). `rand(LKJ(D, 2.0))` returns an array, I don't know how `turing.jl` figures out what the type and size of `P` is; is there a way to tell it that it should be a correlation matrix or at least a `PDMat`? 
Not sure how I would go about "enforc[ing] symmetry there with the same principle"?
<details ><summary>My attempt</summary>

```
using Optim
using StatsPlots
using Turing
using LinearAlgebra


D = 8
N = 100
Ψ =  Matrix{Float64}(I * 10.0, D, D)

# Define a simple Normal model with unknown mean and variance.
@model function mvn_model(X, ::Type{T} = Float64) where {T}
  if X === missing
    # Initialize `x` if missing
    X = Matrix{T}(undef, N, D)
  end
  σ = Vector{T}(undef, D)
  μ ~ MvNormal(zeros(Float64, D), Matrix{Float64}(I * 10.0, D, D))
  P = Symmetric(Matrix{Float64}(undef, D, D))
  Ρ ~ LKJ(D, 2.0)
  for d in 1:D
    σ[d] ~ truncated(Normal(0.0, 1.0), 0.0, Inf)
  end
  Σ = Symmetric(diagm(σ) * Ρ * diagm(σ))
  for n in 1:N
      X[n, :] ~ MvNormal(μ, Σ)
  end
  return (X, μ, Σ)
end

# Draw example data set from prior
example_data = mvn_model(missing)()

#  Run sampler, collect results
chn1 = sample(mvn_model(example_data[1]), HMC(0.1, 5), 1000)
chn2 = sample(mvn_model(example_data[1]), NUTS(0.65), 1000)
```

fails with:

```
PosDefException: matrix is not positive definite; Cholesky factorization failed.
checkpositivedefinite at factorization.jl:18 [inlined]
cholesky!(::Hermitian{ForwardDiff.Dual{ForwardDiff.Tag{Turing.Core.var"#f#1"{DynamicPPL.VarInfo{NamedTuple{(:μ, :Ρ, :σ),Tuple{DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:μ,Tuple{}},Int64},Array{MvNormal{Float64,PDMat{Float64,Array{Float64,2}},Array{Float64,1}},1},Array{DynamicPPL.VarName{:μ,Tuple{}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}},DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:Ρ,Tuple{}},Int64},Array{LKJ{Float64,Int64},1},Array{DynamicPPL.VarName{:Ρ,Tuple{}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}},DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:σ,Tuple{Tuple{Int64}}},Int64},Array{Truncated{Normal{Float64},Continuous,Float64},1},Array{DynamicPPL.VarName{:σ,Tuple{Tuple{Int64}}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}}}},Float64},DynamicPPL.Model{var"#45#46",(:X, :T),(:T,),(),Tuple{Array{Float64,2},Type{Float64}},Tuple{Type{Float64}}},DynamicPPL.Sampler{HMC{Turing.Core.ForwardDiffAD{40},(),AdvancedHMC.UnitEuclideanMetric}},DynamicPPL.DefaultContext},Float64},Float64,10},Array{ForwardDiff.Dual{ForwardDiff.Tag{Turing.Core.var"#f#1"{DynamicPPL.VarInfo{NamedTuple{(:μ, :Ρ, :σ),Tuple{DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:μ,Tuple{}},Int64},Array{MvNormal{Float64,PDMat{Float64,Array{Float64,2}},Array{Float64,1}},1},Array{DynamicPPL.VarName{:μ,Tuple{}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}},DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:Ρ,Tuple{}},Int64},Array{LKJ{Float64,Int64},1},Array{DynamicPPL.VarName{:Ρ,Tuple{}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}},DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:σ,Tuple{Tuple{Int64}}},Int64},Array{Truncated{Normal{Float64},Continuous,Float64},1},Array{DynamicPPL.VarName{:σ,Tuple{Tuple{Int64}}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}}}},Float64},DynamicPPL.Model{var"#45#46",(:X, :T),(:T,),(),Tuple{Array{Float64,2},Type{Float64}},Tuple{Type{Float64}}},DynamicPPL.Sampler{HMC{Turing.Core.ForwardDiffAD{40},(),AdvancedHMC.UnitEuclideanMetric}},DynamicPPL.DefaultContext},Float64},Float64,10},2}}, ::Val{false}; check::Bool) at cholesky.jl:219
cholesky! at cholesky.jl:218 [inlined]
cholesky!(::Array{ForwardDiff.Dual{ForwardDiff.Tag{Turing.Core.var"#f#1"{DynamicPPL.VarInfo{NamedTuple{(:μ, :Ρ, :σ),Tuple{DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:μ,Tuple{}},Int64},Array{MvNormal{Float64,PDMat{Float64,Array{Float64,2}},Array{Float64,1}},1},Array{DynamicPPL.VarName{:μ,Tuple{}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}},DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:Ρ,Tuple{}},Int64},Array{LKJ{Float64,Int64},1},Array{DynamicPPL.VarName{:Ρ,Tuple{}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}},DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:σ,Tuple{Tuple{Int64}}},Int64},Array{Truncated{Normal{Float64},Continuous,Float64},1},Array{DynamicPPL.VarName{:σ,Tuple{Tuple{Int64}}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}}}},Float64},DynamicPPL.Model{var"#45#46",(:X, :T),(:T,),(),Tuple{Array{Float64,2},Type{Float64}},Tuple{Type{Float64}}},DynamicPPL.Sampler{HMC{Turing.Core.ForwardDiffAD{40},(),AdvancedHMC.UnitEuclideanMetric}},DynamicPPL.DefaultContext},Float64},Float64,10},2}, ::Val{false}; check::Bool) at cholesky.jl:251
cholesky! at cholesky.jl:246 [inlined]
cholesky! at cholesky.jl:246 [inlined]
#cholesky#130 at cholesky.jl:344 [inlined]
cholesky at cholesky.jl:344 [inlined]
cholesky at cholesky.jl:344 [inlined]
(::Bijectors.CorrBijector)(::Array{ForwardDiff.Dual{ForwardDiff.Tag{Turing.Core.var"#f#1"{DynamicPPL.VarInfo{NamedTuple{(:μ, :Ρ, :σ),Tuple{DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:μ,Tuple{}},Int64},Array{MvNormal{Float64,PDMat{Float64,Array{Float64,2}},Array{Float64,1}},1},Array{DynamicPPL.VarName{:μ,Tuple{}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}},DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:Ρ,Tuple{}},Int64},Array{LKJ{Float64,Int64},1},Array{DynamicPPL.VarName{:Ρ,Tuple{}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}},DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:σ,Tuple{Tuple{Int64}}},Int64},Array{Truncated{Normal{Float64},Continuous,Float64},1},Array{DynamicPPL.VarName{:σ,Tuple{Tuple{Int64}}},1},Array{Float64,1},Array{Set{DynamicPPL.Selector},1}}}},Float64},DynamicPPL.Model{var"#45#46",(:X, :T),(:T,),(),Tuple{Array{Float64,2},Type{Float64}},Tuple{Type{Float64}}},DynamicPPL.Sampler{HMC{Turing.Core.ForwardDiffAD{40},(),AdvancedHMC.UnitEuclideanMetric}},DynamicPPL.DefaultContext},Float64},Float64,10},2}) at corr.jl:67
logabsdetjac at corr.jl:101 [inlined]
logpdf_with_trans(::LKJ{Float64,Int64}, ::Array{ForwardDiff.Dual{ForwardDiff.Tag{Turing.Core.var"#f#1"{DynamicPPL.VarInfo{NamedTuple{(:μ, :Ρ, :σ),Tuple{DynamicPPL.Metadata{Dict{DynamicPPL.VarName{:μ,Tuple{}},Int64},Array{MvNormal{Float64,PDMat{Float64,Array{Float64,2}},Array{Float64,1}},1},Array{DynamicPPL.VarName{:...
```
</details>

**Reply by devmotion** [reply]
You can't enforce it in this way. Distributions defines the samples of `LKJ` to be dense matrices. This is not a problem per se although it would be better to have an additional distribution `LKJCholesky` that works with the Cholesky factors directly (https://github.com/TuringLang/Bijectors.jl/issues/134#issuecomment-679096131).  I assume that already some of the possible improvements in https://github.com/TuringLang/Bijectors.jl/issues/134#issue-684460340 could be helpful. Unfortunately, the current implementation of the bijector (that reparameterizes the correlation matrices in an unconstrained space that it is easier to work with, in particular for HMC and NUTS) contains quite many weird hacks and workarounds to make sure all AD backends are supported. Maybe your problems could already be improved/solved by replacing https://github.com/TuringLang/Bijectors.jl/blob/b099d4782f675812fe7c67a6f2eace4862d050ae/src/bijectors/corr.jl#L67 with `....cholesky(Symmetric(x))....`

**Reply by jan-glx** [reply]
Thanks for your comment, David. I see that you discussed what I think is the same problem already TuringLang/Bijectors.jl#108 but I can't see if/how it has been solved. In the unlikely case I understand this correctly, wouldn't "the" solution be if `Distributions` defined `LKJ` as a `CorrelationMatrixDistribution`  (inheriting from `ContinuousMatrixDistribution` and being `CorrelationMatrixVariate`) ?

It seems I'd need to learn quite bit more about Julia in general in order to use Turing.jl ... anyway thanks for the help!

**Comment by wetlandscapes** [comment]
Hi,
First, thanks for the _great_ package.

I just come across this issue, as well. I've noticed two things:

1. `LKJ()` does not always produce positive definite matrices. This seems to be checked in the `{LinearAlgebra}` package (`LinearAlgebra.checkpositivedefinite()`). I'm a little shaky on the math, but I think this might be an overzealous error. At least for Cholesky factorized matrices, the requirement is just to be positive semi-definite: https://en.wikipedia.org/wiki/Cholesky_decomposition#Proof_for_positive_semi-definite_matrices. But perhaps that's not the issue (and maybe should be brought up with those maintaining `{LinearAlgebra}`).
2. I sometimes get a domain error when one of the values is exactly -1. My guess is this is another overzealous error, since -1 is clearly fine as a value in correlation matrix. I get this one only rarely, so haven't tracked it down, yet.

The model below is just a toy model I've been playing around with to try and better understand how to use Julia and Turing. The intent was to estimate the means, standard deviations, and correlations from a distribution of known values.

```julia
using Random, Distributions, Turing
using FillArrays, LinearAlgebra
Random.seed!(13)

#Make the data
cols = 3
N = 50
mu = rand(Normal(0, 100), cols)
sd = rand(InverseGamma(3, 10), cols)
cor = rand(LKJ(cols, 1), 1)[1]
cov = sd * sd' .* cor
m = convert(Matrix, rand(MvNormal(mu, cov), N)')

#Use data to estimate priors
mu_p = [mean(col) for col in eachcol(m)]
sd_p = [std(col) for col in eachcol(m)]

@model function correlation_model(data, mu, sd)
  N = size(data)[1]
  C = size(data)[2]
  μ ~ MvNormal(mu, Diagonal(sd))
  #This is not efficient
  σ = fill(undef, length(sd))
  for i in 1:C
    σ[i] ~ Exponential(1 / sd[i])
  end
  #PROBLEM: Sometimes generates a domain error when rho is -1
  #PROBLEM: Sometimes generates a matrix that is not positive definite
    #Note: LinearAlgebra.checkpositivedefinite()
  ρ ~ LKJ(length(σ), 1)
  Σ = σ * σ' .* ρ
  #This is not efficient
  for i in 1:N
    data[i,:] ~ MvNormal(μ, Σ)
  end
end

model = correlation_model(m, mu_p, sd_p)
chain = sample(model, Turing.NUTS(500, 0.65), 1000)
```

I'm really new to Julia (trying to translate my Stan skills... or lack thereof), so I'm very open to the idea that I've done something inappropriate, but hopefully not.

**Comment by joshualeond** [comment]
Hey @jan-glx, I just wanted to bump your discussion since it appears the `LKJCholesky` distribution should now be supported by `Turing.jl`:
[Roadmap to LKJCholesky](https://github.com/TuringLang/Turing.jl/issues/1870)

I'm personally still having problems with it and I'm suspecting it has to do with a difference between `Turing.jl` and Stan in that Stan will reject the current step if there are errors like non-positive definiteness and return an infinite log-density and for `Turing.jl` at least with the NUTS and HMC samplers it will break the sampling altogether with an error.

There are ways to catch these errors and apply the infinite log density yourself (`Turing.@addlogprob! Inf`) but my lack of knowledge on matrix algebra and I suppose numerical linear algebra is making this quite challenging for me to figure out.
